Issue #1352 - Show saved nameservers if registry is down

dave-kennedy-ecs opened the pull request:
## Ticket

Resolves #1352 
<!--Reminder, when a code change is made that is user facing, beyond content updates, then the following are required:
- a developer approves the PR
- a designer approves the PR or checks off all relevant items in this checklist.

All other changes require just a single approving review.-->

## Changes

<!-- What was added, updated, or removed in this PR. -->
- DNS Nameservers (host names and IPs) are stored in the registrar database when they are fetched from registry through EPP
- When registry is unavailable through EPP, DNS Nameservers are retrieved from the registrar database rather than from the registry through EPP

<!--
    Please add/remove/edit any of the template below to fit the needs
    of this specific PR.
--->

## Context for reviewers

<!--Background context, more in-depth details of the implementation, and anything else you'd like to call out or ask reviewers.  -->

## Setup

<!--  Add any steps or code to run in this section to help others run your code.
    
    Example 1:
    ```sh
    echo "Code goes here"
    ``` 
    
    Example 2: If the PR was to add a new link with a redirect, this section could simply be:
    -go to /path/to/start/page
    -click the blue link in the <insert location>
    -notice user is redirected to <proper end location>
-->

## Code Review Verification Steps

### As the original developer, I have

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

### As a code reviewer, I have

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

### As a designer reviewer, I have

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

## Screenshots

<!-- If this PR makes visible interface changes, an image of the finished interface can help reviewers
and casual observers understand the context of the changes.
A before image is optional and can be included at the submitter's discretion.

Consider using an animated image to show an entire workflow.
You may want to use [GIPHY Capture](https://giphy.com/apps/giphycapture) for this! 📸

_Please frame images to show useful context but also highlight the affected regions._
--->

therealslimhsiehdy commented on src/registrar/models/domain.py:
```diff
@@ -1666,6 +1670,51 @@ def _update_hosts_and_contacts(self, cleaned, fetch_hosts, fetch_contacts):
+            # Get all existing hosts from the database for this domain
+            existing_hosts_in_db = Host.objects.filter(domain=self)
+            # Identify hosts to delete
+            cleaned_host_names = set(cleaned_host["name"] for cleaned_host in cleaned_hosts)
```
**Question:** Would you mind giving an example of when we would have duplicates of a host name? (Just trying to think through all edge cases)

dave-kennedy-ecs replied:
There should not be a case of duplicates.  It is not allowed by the logic in the name server form; throws an error 'Remove duplicate entry'.  And it is not allowed by the registry.

therealslimhsiehdy reviewed (commented):
**On local:** Bc EPP doesn't work on local I was able to test out the scenario of retrieving from the registrar DB (via Django Admin > Host > Add new host) rather than from the registry through EPP
<img width="1173" alt="Screenshot 2023-12-28 at 4 02 19 PM" src="https://github.com/cisagov/manage.get.gov/assets/13954664/afddd891-193e-497f-86db-2dda21a622dc">

**On sandbox:** When I add in Host(s) via Django Admin and then make changes via the UI to these Hosts + IPs, I see the changes reflected in the Django Admin -- this means that the database is being updated alongside
<img width="858" alt="Screenshot 2023-12-28 at 4 35 44 PM" src="https://github.com/cisagov/manage.get.gov/assets/13954664/829bdb9f-e9c5-4317-a788-fbcffd96bec5">
<img width="833" alt="Screenshot 2023-12-28 at 4 35 54 PM" src="https://github.com/cisagov/manage.get.gov/assets/13954664/c8b07fc0-8693-4e0f-892f-7dfb8f14f97c">

**Question:** Still milling about the edge case of adding 1 in via Hosts in Django Admin and another via UI and it having the section bork where I can't even add in new Nameservers and IPs even when I deleted it from Hosts. 

Even if we set it it as readonly on Django Admin, wouldn't we still need to apply certain permission settings to add in Hosts (ie make sure they put 2 Nameservers in for 1 domain, but I feel like this may be tricky) to safeguard against unwanted data errors even for an Admin permission level user?

rachidatecs commented:
The doctstrings for the Host and HostIp classes need updating. They still say: "This model exists ONLY to allow a new registrant to draft DNS entries before their application is approved."

dave-kennedy-ecs commented:
> The doctstrings for the Host and HostIp classes need updating. They still say: "This model exists ONLY to allow a new registrant to draft DNS entries before their application is approved."

Updated the docstrings

dave-kennedy-ecs commented:
> **On local:** Bc EPP doesn't work on local I was able to test out the scenario of retrieving from the registrar DB (via Django Admin > Host > Add new host) rather than from the registry through EPP <img alt="Screenshot 2023-12-28 at 4 02 19 PM" width="1173" src="https://private-user-images.githubusercontent.com/13954664/293304000-afddd891-193e-497f-86db-2dda21a622dc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDM4NDg2OTAsIm5iZiI6MTcwMzg0ODM5MCwicGF0aCI6Ii8xMzk1NDY2NC8yOTMzMDQwMDAtYWZkZGQ4OTEtMTkzZS00OTdmLTg2ZGItMmRkYTIxYTYyMmRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyMzEyMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI5VDExMTMxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVlMjNhNzA3YTE0MWMyNGE0NjM1NGJkZDVhODI0ZDg2M2Q3Y2U2M2M1MDhlMTY0MWYyNjkxOTY2YWUyOWRlYTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DwsdHT0B2IcvSiE1yrkkYL0mJL55vR0bDlAmXSTOC1A">
> 
> **On sandbox:** When I add in Host(s) via Django Admin and then make changes via the UI to these Hosts + IPs, I see the changes reflected in the Django Admin -- this means that the database is being updated alongside <img alt="Screenshot 2023-12-28 at 4 35 44 PM" width="858" src="https://private-user-images.githubusercontent.com/13954664/293304038-829bdb9f-e9c5-4317-a788-fbcffd96bec5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDM4NDg2OTAsIm5iZiI6MTcwMzg0ODM5MCwicGF0aCI6Ii8xMzk1NDY2NC8yOTMzMDQwMzgtODI5YmRiOWYtZTljNS00MzE3LWE3ODgtZmJjZmZkOTZiZWM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyMzEyMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI5VDExMTMxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVmZjE4ZDc0YjQyMjBlYzQ2YWM3MzUyODM0YTc3NWVmZTg2NDZlZjIxODNkYWYyNDRiMmRhMDc0YzFmNzQwZjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ybeiq6np1CA9Y3nCKK-rDojCwyzcnO7ASigtwGMypHA"> <img alt="Screenshot 2023-12-28 at 4 35 54 PM" width="833" src="https://private-user-images.githubusercontent.com/13954664/293304043-c8b07fc0-8693-4e0f-892f-7dfb8f14f97c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDM4NDg2OTAsIm5iZiI6MTcwMzg0ODM5MCwicGF0aCI6Ii8xMzk1NDY2NC8yOTMzMDQwNDMtYzhiMDdmYzAtODY5My00ZTBmLTg5MmYtN2RmYjhmMTRmOTdjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyMzEyMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI5VDExMTMxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThiNWY5NGI3OWVlN2NiMTc3ZDkwNTBlMjExZmQxMzFjOTEyNGU4ODkzZmQ3YjZjNGIwNjU1YWQxMzMzMjNmNWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.s8t5j84kjQtmcyhWZ5MyprEPzaxU4y9oItU8pGz_mhw">
> 
> **Question:** Still milling about the edge case of adding 1 in via Hosts in Django Admin and another via UI and it having the section bork where I can't even add in new Nameservers and IPs even when I deleted it from Hosts.
> 
> Even if we set it it as readonly on Django Admin, wouldn't we still need to apply certain permission settings to add in Hosts (ie make sure they put 2 Nameservers in for 1 domain, but I feel like this may be tricky) to safeguard against unwanted data errors even for an Admin permission level user?

I have removed Hosts from django admin.  For two reasons:
1 - any changes made in django admin will never be propagated to registry, so allowing admins to update this data in django admin would be misleading
2 - the logic which enforces rules about valid name servers is not applied in django admin and allows for data which can break functionality for domain management in the app

rachidatecs commented on src/registrar/models/domain.py:
```diff
@@ -1666,6 +1670,51 @@ def _update_hosts_and_contacts(self, cleaned, fetch_hosts, fetch_contacts):
+                [{"name": "ns1.example.com",}, {"name": "ns1.example.gov"}, "addrs": ["0.0.0.0"])]
+            fetch_hosts: boolean indicating whether or not fetch_hosts was called
+        """
+        if fetch_hosts:
```
What do you think about putting this guy outside the method?

therealslimhsiehdy replied:
+1 on this, but definitely non-blocking!

dave-kennedy-ecs replied:
I agree.  Done

rachidatecs commented on src/registrar/models/domain.py:
```diff
@@ -1651,7 +1655,7 @@ def _get_dnssec_data(self, response_extensions):
 
     def _update_hosts_and_contacts(self, cleaned, fetch_hosts, fetch_contacts):
-        """Capture and store old hosts and contacts from cache if they don't exist"""
+        """Capture and cache old hosts and contacts from cache if they don't exist in cleaned"""
```
"""Update cache from old cache"""

Is this what the method's doing?

dave-kennedy-ecs replied:
You're right, I am updating the comment before merging
